src/core/ext/filters/client_channel/resolver/dns/event_engine/event_engine_client_channel_resolver.cc

```diff
@@ -272,24 +270,27 @@ void EventEngineClientChannelDNSResolver::EventEngineDNSRequestWrapper::
   {
     MutexLock lock(&on_resolved_mu_);
     orphaned_ = true;
-    // Event if cancellation fails here, OnResolvedLocked will return early, and
-    // the resolver will never see a completed request.
-    if (hostname_handle_.has_value()) {
-      event_engine_resolver_->CancelLookup(*hostname_handle_);
-    }
-    if (srv_handle_.has_value()) {
-      event_engine_resolver_->CancelLookup(*srv_handle_);
-    }
-    for (const auto& handle : balancer_hostname_handles_) {
-      event_engine_resolver_->CancelLookup(handle);
-    }
-    if (txt_handle_.has_value()) {
-      event_engine_resolver_->CancelLookup(*txt_handle_);
+    if (timeout_handle_.has_value()) {
+      resolver_->event_engine_->Cancel(*timeout_handle_);
+      timeout_handle_.reset();
     }
+    // Even if cancellation fails here, OnResolvedLocked will return early, and
+    // the resolver will never see a completed request.
+    event_engine_resolver_.reset();
   }
   Unref(DEBUG_LOCATION, "Orphan");
 }
 
+void EventEngineClientChannelDNSResolver::EventEngineDNSRequestWrapper::
+    OnTimeout() {
+  MutexLock lock(&on_resolved_mu_);
+  GRPC_EVENT_ENGINE_RESOLVER_TRACE("DNSResolver::%p OnTimeout",
+                                   resolver_.get());
+  timed_out_ = true;
```

Comment:
I believe this bool is redundant, since `timeout_handle_.has_value()` gives the same information (conflated with orphaned). Since `timed_out_` is only used to determine whether to do the subsequent SRV->A record lookups, and we don't want to do those lookups when either orphaned or timed out, I think the bool is unnecessary.

Reply:
Great catch! Yes indeed. We don't want to do the balancer lookups when the SRV lookup completion races with cancellation due to timeout. Here we can check for `!timeout_handle_.has_value()`.